xcode/Subconscious/Shared/Services/UserProfileService.swift

```diff
@@ -344,73 +356,163 @@ actor UserProfileService {
             logger.warning("Failed to sync after updating profile: \(error.localizedDescription)")
         }
     }
-    
-    func loadOurProfileFromMemo() async throws -> UserProfile {
-        let did = try await noosphere.identity()
-        let cid = try await noosphere.version()
+
+    private func readOurProfile(alias: Petname?) async throws -> UserProfile {
+        let identity = try await self.noosphere.identity()
         
         return try await self.loadProfileFromMemo(
-            did: did,
+            did: identity,
             address: Slashlink.ourProfile,
-            resolutionStatus: .resolved(cid)
+            alias: alias
         )
     }
     
-    nonisolated func loadOurProfileFromMemoPublisher(
-    ) -> AnyPublisher<UserProfile, Error> {
-        Future.detached {
-            try await self.loadOurProfileFromMemo()
+    /// Build a UserProfile suitable for list views, transcludes etc.
+    /// This will attempt to read from the database and also maintain an in-memory cache of profiles
+    /// we have encountered.
+    func identifyUser(
+        entry: AddressBookEntry,
+        context: Peer?
+    ) async throws -> UserProfile {
+        return try await self.identifyUser(
+            did: entry.did,
+            petname: entry.petname,
+            context: context
+        )
+    }
+    
+    /// Build a UserProfile suitable for list views, transcludes etc.
+    /// This will attempt to read from the database and also maintain an in-memory cache of profiles
+    /// we have encountered.
+    func identifyUser(
+        did: Did,
+        address: Slashlink,
+        context: Peer?
+    ) async throws -> UserProfile {
+        switch address.peer {
+        case .petname(let petname):
+            return try await self.identifyUser(did: did, petname: petname, context: context)
+        case .did, .none:
+            return try await self.identifyUser(did: did, petname: nil, context: context)
         }
-        .eraseToAnyPublisher()
     }
     
-    func buildUserProfile(
-        address: Slashlink
+    /// Build a UserProfile suitable for list views, transcludes etc.
+    /// This will attempt to read from the database and also maintain an in-memory cache of profiles
+    /// we have encountered.
+    ///
+    /// `context` will be used to determine the preferred navigation address for a user.
+    func identifyUser(
```

Comment:
There's a lot going on in this function. Probably deserves a test.

Reply:
I had thought this would be too difficult to test without traversal, but I came up with some reasonable test cases that cover the logic within this function itself. 

See `Tests_UserProfileService.testIdentifyUserCallSignatures` and `Tests_UserProfileService.testIdentifyUserRebaseOnContext`